File: webOS-legacy/jukie-drm/internal/appleauth/webtoken.go
// Package appleauth obtains Apple Music credentials for Jukie.
//
// webtoken.go: fetch the PUBLIC first-party "web developer token" (the ~35-day
// AMPWebPlay token the music.apple.com player uses). This needs NO login - the
// token is baked into the site - so it can be auto-refreshed silently. jukie-drm
// uses it as the Bearer for play.music.apple.com webPlayback.
//
// The device's old curl can't TLS-handshake music.apple.com, but Go's HTTP client
// can (same as the existing play.music.apple.com calls), so this lives in Go.
package appleauth

import (
	"encoding/base64"
	"encoding/json"
	"fmt"
	"io"
	"net/http"
	"net/url"
	"os"
	"path/filepath"
	"regexp"
	"strings"
	"time"
)

// DefaultCacheDir is the jail-writable persistent dir on the TouchPad.
const DefaultCacheDir = "/media/internal/.jukie"

// Logf is an optional progress sink (set by the CLI to write to stderr). No-op by default.
var Logf = func(string, ...interface{}) {}

const browserUA = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36"

// Pages that carry the web player bootstrap (and thus the token).
var tokenPages = []string{
	"https://music.apple.com/us/browse",
	"https://music.apple.com/",
}

var (
	jwtRe       = regexp.MustCompile(`eyJ[A-Za-z0-9_-]{8,}\.[A-Za-z0-9_-]{20,}\.[A-Za-z0-9_-]{8,}`)
	bundleRe    = regexp.MustCompile(`/assets/[A-Za-z0-9_./-]*index[A-Za-z0-9_.-]*\.js`)
	scriptSrcRe = regexp.MustCompile(`<script[^>]+src="([^"]+\.js)"`)
)

// WebToken is a fetched first-party web developer token and its expiry.
type WebToken struct {
	Token  string    `json:"token"`
	Expiry time.Time `json:"expiry"`
}

// FetchWebToken scrapes music.apple.com for the first-party developer token. No auth.
func FetchWebToken(client *http.Client) (WebToken, error) {
	if client == nil {
		client = &http.Client{Timeout: 20 * time.Second}
	}
	var lastErr error
	for _, page := range tokenPages {
		Logf("GET %s", page)
		html, err := httpGet(client, page)
		if err != nil {
			Logf("  err: %v", err)
			lastErr = err
			continue
		}
		Logf("  %d bytes", len(html))
		// 1) token may be inline in the HTML (meta config or a raw JWT)
		if t, ok := validToken(findTokenInText(string(html))); ok {
			Logf("  found token inline in HTML")
			return t, nil
		}
		// 2) otherwise pull the JS bundles and scan them
		bundles := bundleURLs(string(html))
		Logf("  scanning %d JS bundle(s)", len(bundles))
		for _, u := range bundles {
			js, err := httpGet(client, u)
			if err != nil {
				Logf("    %s err: %v", u, err)
				continue
			}
			Logf("    %s (%d bytes)", u, len(js))
			if t, ok := validToken(findTokenInText(string(js))); ok {
				Logf("    found token in bundle")
				return t, nil
			}
		}
	}
	if lastErr != nil {
		return WebToken{}, fmt.Errorf("fetch music.apple.com: %w", lastErr)
	}
	return WebToken{}, fmt.Errorf("no usable developer token found on music.apple.com")
}

func httpGet(client *http.Client, u string) ([]byte, error) {
	req, _ := http.NewRequest("GET", u, nil)
	req.Header.Set("User-Agent", browserUA)
	req.Header.Set("Accept", "text/html,application/xhtml+xml,application/javascript,*/*")
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	defer resp.Body.Close()
	if resp.StatusCode != 200 {
		return nil, fmt.Errorf("HTTP %d for %s", resp.StatusCode, u)
	}
	return io.ReadAll(io.LimitReader(resp.Body, 12<<20)) // 12MB cap
}

// findTokenInText returns the first JWT that looks like the web dev token
// (has root_https_origin + exp), else any JWT carrying an exp.
func findTokenInText(s string) string {
	if dec, err := url.QueryUnescape(s); err == nil && dec != s {
		s = s + "\n" + dec // meta config is URL-encoded JSON
	}
	var fallback string
	for _, m := range jwtRe.FindAllString(s, -1) {
		if isWebDevToken(m) {
			return m
		}
		if fallback == "" && jwtExp(m) != nil {
			fallback = m
		}
	}
	return fallback
}

func bundleURLs(html string) []string {
	var out []string
	seen := map[string]bool{}
	add := func(p string) {
		if p == "" || seen[p] || !strings.HasSuffix(p, ".js") {
			return
		}
		seen[p] = true
		switch {
		case strings.HasPrefix(p, "http"):
			out = append(out, p)
		case strings.HasPrefix(p, "/"):
			out = append(out, "https://music.apple.com"+p)
		}
	}
	for _, m := range bundleRe.FindAllString(html, -1) {
		add(m)
	}
	for _, m := range scriptSrcRe.FindAllStringSubmatch(html, -1) {
		add(m[1])
	}
	if len(out) > 6 {
		out = out[:6]
	}
	return out
}

// ---- JWT helpers ----

func jwtPayload(tok string) map[string]interface{} {
	parts := strings.Split(tok, ".")
	if len(parts) != 3 {
		return nil
	}
	raw, err := base64.RawURLEncoding.DecodeString(parts[1])
	if err != nil {
		return nil
	}
	var m map[string]interface{}
	if json.Unmarshal(raw, &m) != nil {
		return nil
	}
	return m
}

func jwtExp(tok string) *time.Time {
	m := jwtPayload(tok)
	if m == nil {
		return nil
	}
	if e, ok := m["exp"].(float64); ok {
		t := time.Unix(int64(e), 0)
		return &t
	}
	return nil
}

func isWebDevToken(tok string) bool {
	m := jwtPayload(tok)
	if m == nil {
		return false
	}
	_, hasOrigin := m["root_https_origin"]
	_, hasExp := m["exp"]
	return hasOrigin && hasExp
}

func validToken(tok string) (WebToken, bool) {
	if tok == "" {
		return WebToken{}, false
	}
	exp := jwtExp(tok)
	if exp == nil {
		return WebToken{}, false
	}
	return WebToken{Token: tok, Expiry: *exp}, true
}

// ---- cache + ensure ----

func webTokenCachePath(dir string) string {
	if dir == "" {
		dir = DefaultCacheDir
	}
	return filepath.Join(dir, "webtoken.cache.json")
}

// LoadCachedWebToken returns a cached token if present and not within a day of expiry.
func LoadCachedWebToken(dir string) (WebToken, bool) {
	b, err := os.ReadFile(webTokenCachePath(dir))
	if err != nil {
		return WebToken{}, false
	}
	var t WebToken
	if json.Unmarshal(b, &t) != nil || t.Token == "" {
		return WebToken{}, false
	}
	if time.Until(t.Expiry) < 24*time.Hour {
		return WebToken{}, false
	}
	return t, true
}

// SaveCachedWebToken writes the token to the cache dir (best-effort).
func SaveCachedWebToken(dir string, t WebToken) error {
	if dir == "" {
		dir = DefaultCacheDir
	}
	if err := os.MkdirAll(dir, 0777); err != nil {
		return err
	}
	b, _ := json.MarshalIndent(t, "", "  ")
	return os.WriteFile(webTokenCachePath(dir), b, 0666)
}

// EnsureWebToken returns a valid web token: prefer `current` if still valid, else the
// cache, else a fresh fetch (which it caches). Falls back to `current` on fetch failure.
func EnsureWebToken(dir string, client *http.Client, current string) (string, error) {
	if tokenStillValid(current) {
		return current, nil
	}
	if t, ok := LoadCachedWebToken(dir); ok {
		return t.Token, nil
	}
	t, err := FetchWebToken(client)
	if err != nil {
		if current != "" {
			return current, nil // better stale than nothing
		}
		return "", err
	}
	_ = SaveCachedWebToken(dir, t)
	return t.Token, nil
}

func tokenStillValid(tok string) bool {
	if tok == "" {
		return false
	}
	exp := jwtExp(tok)
	return exp != nil && time.Until(*exp) > 24*time.Hour
}

File: webOS-legacy/com.achunt.jukie/enyo/utility/webtoken.js
/*globals window, fetch, Promise, WebToken:true */
/*
 * WebToken - client-side port of jukie-drm's internal/appleauth/webtoken.go: scrapes
 * music.apple.com for the PUBLIC "AMPWebPlay" web-player developer token (~35-day JWT,
 * no login needed - it's baked into the site for anyone visiting). jukie-drm already
 * auto-refreshes its OWN copy of this token for full-track playback; this app's own
 * catalog/library browsing (AppleMusicService) had no equivalent, so a manually-pasted
 * Settings token going stale silently broke "Your Library" - see [[settings-save-db8-queue-bug]]
 * memory / the 2026-07-27 session that found the 401.
 *
 * CONFIRMED WORKING on-device (2026-07-27), somewhat surprisingly: jukie-drm's own fetch
 * runs in a separate Go process with no browser-style cross-origin restrictions, but this
 * app's fetch() (utility/fetch-polyfill.js, backed by XMLHttpRequest) had no such
 * guarantee - a modern browser would block reading a cross-origin response body without
 * matching CORS headers, which music.apple.com's page/bundle responses don't send. This
 * old webOS 3.x WebKit apparently doesn't enforce that restriction: a forced end-to-end
 * test (GET /us/browse -> scan its JS bundles -> find + apply a real, currently-valid
 * AMPWebPlay token) succeeded, and the token verified byte-identical (by hash) to one
 * independently scraped straight from Apple's live site outside the app. If Apple or a
 * future webOS/WebAppMgr patch ever changes this, every call here just fails
 * (rejects/logs) and callers fall back to whatever token they already had - no worse off
 * than before this file existed.
 */
(function (global) {
    "use strict";

    var UA = "Mozilla/5.0 (Windows NT 10.0; Win64; x64) AppleWebKit/537.36 (KHTML, like Gecko) Chrome/120.0.0.0 Safari/537.36";
    // Kept short: this can run reactively off a live 401 (see applemusicservice.js's
    // request()), and worst case (every fetch times out rather than failing fast) is
    // roughly TOKEN_PAGES.length * (1 + MAX_BUNDLES_TO_SCAN) * FETCH_TIMEOUT_MS before
    // giving up - a real user-visible stall if set too generously.
    var FETCH_TIMEOUT_MS = 8000;
    var MAX_BUNDLES_TO_SCAN = 2;
    var REFRESH_MARGIN_MS = 24 * 60 * 60 * 1000; // refresh a day before actual expiry, matching jukie-drm

    var TOKEN_PAGES = ["https://music.apple.com/us/browse", "https://music.apple.com/"];
    var JWT_RE = /eyJ[A-Za-z0-9_-]{8,}\.[A-Za-z0-9_-]{20,}\.[A-Za-z0-9_-]{8,}/g;
    var BUNDLE_RE = /\/assets\/[A-Za-z0-9_./-]*index[A-Za-z0-9_.-]*\.js/g;
    var SCRIPT_SRC_RE = /<script[^>]+src="([^"]+\.js)"/g;

    // Builds ONE string and makes a single console.log call, rather than
    // console.log.apply(console, [multiple, separate, args]) - confirmed on-device that
    // webOS's console->syslog bridge silently drops every argument past the first when
    // called that way (matches the pattern kindSettings.log already uses for the same
    // reason - see settings.js).
    function log() {
        var parts = ["[WebToken]"];
        var i;
        if (typeof console === "undefined" || !console.log) {
            return;
        }
        for (i = 0; i < arguments.length; i++) {
            parts.push((typeof arguments[i] === "object") ? JSON.stringify(arguments[i]) : ("" + arguments[i]));
        }
        console.log(parts.join(" "));
    }

    function b64urlDecode(s) {
        var padded = s + new Array((4 - (s.length % 4)) % 4 + 1).join("=");
        var normal = padded.replace(/-/g, "+").replace(/_/g, "/");
        return decodeURIComponent(escape(window.atob(normal)));
    }

    function jwtPayload(tok) {
        var parts = (tok || "").split(".");
        if (parts.length !== 3) {
            return null;
        }
        try {
            return JSON.parse(b64urlDecode(parts[1]));
        } catch (err) {
            return null;
        }
    }

    function jwtExpiry(tok) {
        var p = jwtPayload(tok);
        if (!p || typeof p.exp !== "number") {
            return null;
        }
        return new Date(p.exp * 1000);
    }

    function isWebDevToken(tok) {
        var p = jwtPayload(tok);
        return !!(p && p.root_https_origin && typeof p.exp === "number");
    }

    // First JWT that looks like the web dev token (root_https_origin + exp), else any
    // JWT carrying an exp - mirrors webtoken.go's findTokenInText.
    function findTokenInText(text) {
        var matches = text.match(JWT_RE) || [];
        var fallback = "";
        var i;
        for (i = 0; i < matches.length; i++) {
            if (isWebDevToken(matches[i])) {
                return matches[i];
            }
            if (!fallback && jwtExpiry(matches[i])) {
                fallback = matches[i];
            }
        }
        return fallback;
    }

    function bundleURLs(html) {
        var out = [];
        var seen = {};
        var m;

        function add(p) {
            var url;
            if (!p || seen[p] || p.indexOf(".js") === -1) {
                return;
            }
            seen[p] = true;
            if (p.indexOf("http") === 0) {
                url = p;
            } else if (p.indexOf("/") === 0) {
                url = "https://music.apple.com" + p;
            } else {
                return;
            }
            out.push(url);
        }

        BUNDLE_RE.lastIndex = 0;
        while ((m = BUNDLE_RE.exec(html))) {
            add(m[0]);
        }
        SCRIPT_SRC_RE.lastIndex = 0;
        while ((m = SCRIPT_SRC_RE.exec(html))) {
            add(m[1]);
        }
        return out.slice(0, MAX_BUNDLES_TO_SCAN);
    }

    function fetchTextWithTimeout(url) {
        return new Promise(function (resolve, reject) {
            var settled = false;
            var timer = setTimeout(function () {
                if (settled) { return; }
                settled = true;
                reject(new Error("timed out: " + url));
            }, FETCH_TIMEOUT_MS);

            fetch(url, { method: "GET", headers: { "Accept": "text/html,application/javascript,*/*", "User-Agent": UA } })
                .then(function (response) {
                    if (settled) { return; }
                    if (!response.ok) {
                        throw new Error("HTTP " + response.status + " for " + url);
                    }
                    return response.text();
                })
                .then(function (text) {
                    if (settled) { return; }
                    settled = true;
                    clearTimeout(timer);
                    resolve(text);
                })
                .catch(function (err) {
                    if (settled) { return; }
                    settled = true;
                    clearTimeout(timer);
                    reject(err);
                });
        });
    }

    // Scrapes music.apple.com for the public web developer token. No auth. Rejects
    // (never throws synchronously) if nothing usable is found or every fetch failed
    // (most likely a cross-origin block - see file header).
    function fetchWebDeveloperToken() {
        var pageIdx = 0;

        function tryPage() {
            if (pageIdx >= TOKEN_PAGES.length) {
                return Promise.reject(new Error("no usable developer token found on music.apple.com"));
            }
            var page = TOKEN_PAGES[pageIdx];
            pageIdx += 1;
            log("GET", page);
            return fetchTextWithTimeout(page).then(function (html) {
                var inline = findTokenInText(html);
                if (inline) {
                    log("found token inline in HTML");
                    return inline;
                }
                var bundles = bundleURLs(html);
                log("scanning", bundles.length, "JS bundle(s)");
                return scanBundles(bundles, 0);
            }).catch(function (err) {
                log("  page failed:", page, err && err.message);
                return tryPage();
            });
        }

        function scanBundles(bundles, idx) {
            if (idx >= bundles.length) {
                return tryPage();
            }
            return fetchTextWithTimeout(bundles[idx]).then(function (js) {
                var tok = findTokenInText(js);
                if (tok) {
                    log("found token in bundle", bundles[idx]);
                    return tok;
                }
                return scanBundles(bundles, idx + 1);
            }).catch(function (err) {
                log("  bundle failed:", bundles[idx], err && err.message);
                return scanBundles(bundles, idx + 1);
            });
        }

        return tryPage().then(function (tok) {
            var expiry = jwtExpiry(tok);
            if (!expiry) {
                throw new Error("scraped token has no exp claim");
            }
            return { token: tok, expiry: expiry };
        });
    }

    function stillValid(tok) {
        var expiry = jwtExpiry(tok);
        return !!(tok && expiry && (expiry.getTime() - Date.now()) > REFRESH_MARGIN_MS);
    }

    // Resolves { token, changed } - `current` unchanged if still valid or if the fetch
    // fails (better stale than nothing, matching webtoken.go's EnsureWebToken). Never
    // rejects, so callers can fire this at boot without extra error handling. Pass
    // `force` to skip the exp-based validity check and re-scrape unconditionally
    // (used on a live 401 - the token's own exp claim can say "fine" while the server
    // disagrees, e.g. revoked).
    function ensure(current, force) {
        if (!force && stillValid(current)) {
            return Promise.resolve({ token: current, changed: false });
        }
        return fetchWebDeveloperToken().then(function (result) {
            if (result.token === current) {
                return { token: current, changed: false };
            }
            return { token: result.token, changed: true };
        }).catch(function (err) {
            log("auto-refresh unavailable (blocked or unreachable), keeping existing token:", err && err.message);
            return { token: current, changed: false };
        });
    }

    global.WebToken = {
        ensure: ensure,
        stillValid: stillValid,
        jwtExpiry: jwtExpiry
    };
}(window));
